### PR TITLE
[tune/rllib] Use cloudpickle to dump config

### PR DESCRIPTION
**Pull request by eugenevinitsky**
## What do these changes do?

JSON Logger now uses cloudpickle to dump the configs as welll, which pkls the functions needed for multi-agent replay.

### CODE REVIEW

File: python/ray/tune/logger.py [hunk 1]
```diff
@@ -3,6 +3,7 @@
 from __future__ import print_function
 
 import csv
+import dill
```
Comment (ericl) [inline python/ray/tune/logger.py]: try:
   import dill
except ImportError:
   dill = None

File: python/ray/tune/logger.py [hunk 2]
```diff
@@ -103,6 +104,11 @@ def _init(self):
                 indent=2,
                 sort_keys=True,
                 cls=_SafeFallbackEncoder)
+        config_pkl = os.path.join(self.logdir, "params.pkl")
```
Comment (ericl) [inline python/ray/tune/logger.py]: if dill != None:
  ....
else:
  logger.info("Not saving pickled config since `dill` is not available.")

**Review by ericl — COMMENTED** [review]
Looks good but let's make the use of dill a soft dependency.

File: python/ray/tune/logger.py (hunk 2 above)
Comment (eugenevinitsky) [reply python/ray/tune/logger.py]: Makes sense, thank you!

File: python/ray/tune/logger.py (hunk 1 above)
Comment (eugenevinitsky) [reply python/ray/tune/logger.py]: Ty!

**Comment by robertnishihara** [comment]
Should we add dill to the rllib `setup.py` dependencies?

Also, why are we using dill?

If dill is not installed, what happens? Do we use a different serialization format or just not use anything?

**Comment by eugenevinitsky** [comment]
Currently, some of the classes and functions (for example the policy graph and tune functions mapping the policy ids to graphs) needed to run a multiagent experiment wind up getting stored as strings, which makes replay difficult. Dill solves this problem.

**Comment by eugenevinitsky** [comment]
No I haven't, I can try this out!

**Comment by robertnishihara** [comment]
@eugenevinitsky if that works then we should definitely use that instead of dill. Ideally we wouldn't have tons of different serialization formats floating around.

Also, if you rely on the installed version of dill (or cloudpickle), then you can run into version incompatibilities on different machines. This is why we ship cloudpickle with Ray.

**Comment by eugenevinitsky** [comment]
@robertnishihara you were right; cloudpickle worked fine.

**Review by ericl — APPROVED** [review]

File: python/ray/tune/logger.py [hunk 3]
```diff
@@ -9,6 +9,7 @@
 import os
 import yaml
 
+from ray.cloudpickle import cloudpickle
```
Comment (robertnishihara) [inline python/ray/tune/logger.py]: Can you change this to `import ray.cloudpickle as pickle` or `import ray.cloudpickle as cloudpickle`.

Comment (eugenevinitsky) [reply python/ray/tune/logger.py]: yeah of course, sorry for the deviation from the standards

**Pull request merged**